### ISSUE TITLE
Loss of precision in Integer -> Float autocast

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#11710]
When adding [number autocast](https://github.com/crystal-lang/crystal/pull/11431) for 1.3.0, one key aspect of the design was to preserve the precision of numbers. That's why it's possible to cast an Int32 to a Float64, but not to Float32. However, literal autocasting might lose precision:

```crystal
a : Float32 = 0
a = 33554435_i32
p a # prints 33554436.0
```

Probably the easiest fix it to see if for FloatN the literal integer fits in an (U)Int(N/2).

**Comment by yxhuvud** [comment]
That could detect the error but what should the action taken be? Compilation error?

**Comment by beta-ziliani** [comment]
Thanks @yxhuvud , I updated the issue.

**Comment by straight-shoota** [comment]
Yes, I think this should be a compilation error. If you don't care about precision, you wouldn't use an integer literal.

We should also consider Float32 to Float64 autocasting. 

Float64's range is much bigger as well. Literals outside of Float32's range silently turn up as `Infinity`. That should certainly be a compiler error instead.

```cr
x : Float32
x = 1.7976931348623157e+308 # => Infinity
```

Additionally, Float64 has higher precision than Float32. The problem with floating point literals is that they are almost always losing precision.
```cr
0.99999999_f32 # => 1.0
0.99999999_f64 # => 0.99999999
0.99999999999999999_f64 # => 1.0
```

So it's clear that a float literal is never really precise. Even that precise looking `0.99999999` is not that exact decimal value, it's just very close to it.
Still, when using a literal explicitly typed as Float64, the expectation is to represent a floating point value with 64-bit precision. Not 32-bits.  So I think an explicit Float64 literal should *never* autocast to Float32.

**Comment by straight-shoota** [comment]
I realized the range problem isn't specific to autocasting, it also shows with regular literals:

```cr
1.7976931348623157e+308_f32 # => Infinity
```
